domain/permission/state/state.go

```diff
@@ -384,3 +551,117 @@ func objectTag(id corepermission.ID) (result names.Tag) {
+FROM permission_access_type
+     JOIN permission ON permission_access_type.id = permission.permission_type_id
+WHERE permission.grant_on = $readUserPermission.grant_on AND permission.grant_to = $readUserPermission.grant_to
+`
+	findAccessTypeStmt, err := st.Prepare(findAccessTypeQuery, readUserPermission{})
+	if err != nil {
+		return "", "", errors.Annotate(err, "preparing select findAccessType query")
+	}
+
+	input := readUserPermission{
+		GrantTo: grantTo,
+		GrantOn: grantOn,
+	}
+	result := readUserPermission{}
+	err = tx.Query(ctx, findAccessTypeStmt, input).Get(&result)
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return "", "", errors.Annotatef(permissionerrors.NotFound, "for %q on %q", grantTo, grantOn)
+	} else if err != nil {
+		return "", "", errors.Annotatef(err, "getting permission for %q on %q", grantTo, grantOn)
+	}
+
+	return result.AccessType, result.UUID, nil
+}
+
+// readUsersPermissions returns all permissions for the grantTo, a user UUID.
+func (st *State) readUsersPermissions(ctx context.Context,
+	tx *sqlair.TX,
+	grantTo string,
+) ([]readUserPermission, error) {
+	query := `
+SELECT (permission.uuid, permission.grant_on) AS (&readUserPermission.*),
+       permission_access_type.type AS &readUserPermission.access_type
+FROM permission
+     JOIN permission_access_type
+     ON permission_access_type.id = permission.permission_type_id
+WHERE permission.grant_to = $M.grant_to
+`
+	// Validate the grant_on target exists.
+	stmt, err := st.Prepare(query, readUserPermission{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var usersPermissions = []readUserPermission{}
+	err = tx.Query(ctx, stmt, sqlair.M{"grant_to": grantTo}).GetAll(&usersPermissions)
+	if err != nil {
+		return nil, errors.Annotatef(err, "collecting permissions for %q", grantTo)
+	}
+
+	if len(usersPermissions) >= 1 {
+		return usersPermissions, nil
+	}
+	return nil, errors.Annotatef(permissionerrors.NotFound, "for %q", grantTo)
+}
+
+func grantOnType(ctx context.Context,
+	tx *sqlair.TX,
+	permissions []readUserPermission,
+) ([]readUserPermission, error) {
+	for _, p := range permissions {
+		// TODO: can we make objectType work on a slice of GrantOn
+		// in both use cases?
+		objectType, err := objectType(ctx, tx, p.GrantOn)
+		if err != nil {
+			return nil, errors.Annotatef(err, "finding type for %q", p.GrantOn)
+		}
+		p.ObjectType = objectType
+	}
+	return permissions, nil
+}
+
+// targetPermissions returns a slice of readUserPermission for
+// every permission available for the given target specified by
+// grantOn.
+func (st *State) targetPermissions(ctx context.Context,
+	tx *sqlair.TX,
+	grantOn string,
+) ([]readUserPermission, error) {
+	query := `
+SELECT (permission.uuid, permission.grant_on, permission.grant_to) AS (&readUserPermission.*),
+       permission_access_type.type AS &readUserPermission.access_type
+FROM permission
+     JOIN permission_access_type
```

Comment:
Third time warrants a view.

Reply:
@manadart The entire thing? The `JOIN permission_access_type` part? It's not obvious based on our current usage of view how to do the small part.  (And being newer to SQL stmts). More detail is appreciated.